### PR TITLE
MemoryManager: fix deallocation bugs

### PR DESCRIPTION
**Pull request by sporksmith**

### CODE REVIEW

**Comment by sporksmith** [comment]
It might be useful to look at the individual commits. I separate it into a small commit with the actual bugfixes, and all the cleanup + extra logging I did while debugging.

**Comment by sporksmith** [comment]
Sorry - pushed one more commit to add regression tests

File: src/test/memory/test_mmap.rs [hunk 1]
```diff
@@ -181,6 +187,48 @@ fn test_mmap_anon() -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
+fn test_munmap_front() -> Result<(), Box<dyn Error>> {
```
Comment (stevenengler) [inline src/test/memory/test_mmap.rs:190]: A short comment for each test would be helpful to explain what's being tested or why.

**Review by stevenengler — APPROVED** [review]

File: src/main/host/memory_manager.rs [hunk 2]
```diff
@@ -450,7 +450,7 @@ impl MemoryManager {
                     // Unmap range from Shadow's address space.
                     unsafe {
                         sys::mman::munmap(
-                            (left_region.shadow_base as usize + left.end) as *mut c_void,
+                            (left_region.shadow_base.add(left.len())) as *mut c_void,
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/host/memory_manager.rs:462]: Good point. Done.

File: src/main/host/memory_manager.rs [hunk 3]
```diff
@@ -893,6 +904,7 @@ impl MemoryManager {
         size: usize,
         prot: i32,
     ) -> nix::Result<()> {
+        debug!("mprotect({:?}, {}, {:?}", addr, size, prot);
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/host/memory_manager.rs]: done

File: src/test/memory/test_mmap.rs (hunk 1 above)
Comment (sporksmith) [reply src/test/memory/test_mmap.rs:192]: Done

**Pull request merged**